userspace/libscap/scap_savefile.c

```diff
@@ -2523,6 +2748,70 @@ int32_t scap_next_offline(scap_t *handle, OUT scap_evt **pevent, OUT uint16_t *p
 		*pevent = (struct ppm_evt_hdr *)(handle->m_file_evt_buf + sizeof(uint16_t));
 	}
 
+	if((*pevent)->type >= PPM_EVENT_MAX)
+	{
+		//
+		// We're reading a capture that contains new syscalls.
+		// We can't do anything else that skips them.
+		//
+		return scap_next_offline(handle, pevent, pcpuid);
+	}
+
+	if(bh.block_type != EV_BLOCK_TYPE_V2 && bh.block_type != EVF_BLOCK_TYPE_V2)
+	{
+		//
+		// We're reading a old capture which events don't have nparams in the header.
+		// Convert it to the current version.
+		//
+		if((readlen + sizeof(uint32_t)) > FILE_READ_BUF_SIZE)
+		{
+			snprintf(handle->m_lasterr, SCAP_LASTERR_SIZE, "cannot convert v1 event block to v2 (%lu greater than read buffer size %u)",
+				 readlen + sizeof(uint32_t),
+				 FILE_READ_BUF_SIZE);
+			return SCAP_FAILURE;
+		}
+	        size_t offset =
+#ifdef PPM_ENABLE_SENTINEL
+			sizeof(uint32_t) + // sentinel
+#endif
+			sizeof(uint64_t) + // ts
+			sizeof(uint64_t) + // tid
+			sizeof(uint32_t) + // len
+			sizeof(uint16_t);  // type
+
+		memmove((char *)*pevent + offset + sizeof(uint32_t),
+			(char *)*pevent + offset,
+			readlen - ((char *)*pevent - handle->m_file_evt_buf) - offset);
+		(*pevent)->len += sizeof(uint32_t);
+
+		//
+		// The number of parameters needs to be calculated based on the block len.
+		// Use the current number of parameters as starting point and decrease it
+		// until size matches.
+		//
+		char *end = (char *)*pevent + (*pevent)->len;
+		uint16_t *lens = (uint16_t *)((char *)*pevent + sizeof(struct ppm_evt_hdr));
+		uint32_t nparams;
+		for(nparams = g_event_info[(*pevent)->type].nparams; nparams >= 0; nparams--)
+		{
+			char *valptr = (char *)lens + nparams * sizeof(uint16_t);
+			if(valptr > end)
+			{
+				break;

```

Comment:
This definitely shouldn't happen, so should we consider throwing an error instead of silently declare success?

Reply:
Actually, the issue there is that it should be a `continue`, not a `break`.
In the loop, `nparams` starts with the __current__ event table value for that type. So, for example, while parsing an event that in the capture has 0 params but in a new version has some, that check is needed to avoid reading garbage or even out of bounds and continue the loop until `nparams` has been decreased to zero.
Makes sense?